transformer_lens/head_detector.py

```diff
@@ -1,158 +1,269 @@
-    remaining divided by the total attention would then be our score for that attention head. (This particular head is already implemented in HEAD_NAMES)
+from transformer_lens.utils import is_lower_triangular, is_square
+
+HeadName = Literal["previous_token_head", "duplicate_token_head", "induction_head"]
+HEAD_NAMES = cast(List[HeadName], get_args(HeadName))
+ErrorMeasure = Literal["abs", "mul"]
+
+LayerHeadTuple = Tuple[int, int]
+LayerToHead = Dict[int, List[int]]
+
+INVALID_HEAD_NAME_ERR = (
+    f"detection_pattern must be a Tensor or one of head names: {HEAD_NAMES}; got %s"
+)
+
+SEQ_LEN_ERR = (
+    "The sequence must be non-empty and must fit within the model's context window."
+)
+
+DET_PAT_NOT_SQUARE_ERR = "The detection pattern must be a lower triangular matrix of shape (sequence_length, sequence_length); sequence_length=%d; got detection patern of shape %s"
+
+
+def detect_head(
+    model: HookedTransformer,
+    seq: Union[str, List[str]],
+    detection_pattern: Union[torch.Tensor, HeadName],
+    heads: Optional[Union[List[LayerHeadTuple], LayerToHead]] = None,
+    cache: Optional[ActivationCache] = None,
+    *,
+    exclude_bos: bool = False,
+    exclude_current_token: bool = False,
+    error_measure: ErrorMeasure = "mul",
+) -> torch.Tensor:
+    """Searches the model (or a set of specific heads, for circuit analysis) for a particular type of attention head.
+    This head is specified by a detection pattern, a (sequence_length, sequence_length) tensor representing the attention pattern we expect that type of attention head to show.
+    The detection pattern can be also passed not as a tensor, but as a name of one of pre-specified types of attention head (see `HeadName` for available patterns), in which case the tensor is computed within the function itself.
+
+    There are two error measures available for quantifying the match between the detection pattern and the actual attention pattern.
+
+    1. `"mul"` (default) multiplies both tensors element-wise and divides the sum of the result by the sum of the attention pattern.
+    Typically, the detection pattern should in this case contain only ones and zeros, which allows a straightforward interpretation of the score:
+    how big fraction of this head's attention is allocated to these specific query-key pairs?
+    Using values other than 0 or 1 is not prohibited but will raise a warning (which can be disabled, of course).
+    2. `"abs"` calculates the mean element-wise absolute difference between the detection pattern and the actual attention pattern.
+    The "raw result" ranges from 0 to 2 where lower score corresponds to greater accuracy. Subtracting it from 1 maps that range to (-1, 1) interval,
+    with 1 being perfect match and -1 perfect mismatch.
+
+    **Which one should you use?** `"abs"` is likely better for quick or exploratory investigations. For precise examinations where you're trying to
```

Comment:
This should be `"mul` is likely better for quick or exploratory investigations.

Reply:
thanks! done